text/0111-auto-resource-detection.md

```diff
@@ -0,0 +1,115 @@
+
+A default implementation of a detector that reads resource data from the `OTEL_RESOURCE` environment variable will be included in the SDK. The environment variable will contain of a list of key value pairs, and these are expected to be represented in a format similar to the [W3C Correlation-Context](https://github.com/w3c/correlation-context/blob/master/correlation_context/HTTP_HEADER_FORMAT.md#header-value), i.e.: `key1=value1,key2=value2`. This detector must always be configured as the first detector by default.
+
+Custom resource detectors related to specific environments (e.g. specific cloud vendors) must be implemented outside of the SDK, and users will need to import these separately.
+
+In order to allow users to be able to use different detectors for different trace or meter providers, the detector(s) must be configured against a resource provider, and users can create multiple providers if desired.
+
+## Internal details
+
+As described above, the following will be added to the Resource SDK specification:
+
+- An interface for "detectors", to retrieve resource information
+- Specification for the resource provider, which can have custom detectors added, and will return a merged resource
+- Details of the default "From Environment Variable" detector implementation described above
+
+This is a relatively small proposal so is easiest to explain the details with a code example:
+
+### Usage
+
+The following example in Go creates a trace provider that uses resource information automatically detected from AWS or GCP:
+
+Assumes a dependency has been added on the `otel api`, `otel sdk`, `otel awsdetector`, and `otel gcpdetector` packages.
+
+```go
+rp := resource.NewMustProvider() // or NewProvider (see below)
+rp.AddDetectors(awsdetector.EC2, gcpdetector.GCE, gcpdetector.GKE)
+resource := rp.AutoDetect(ctx)
+tp := sdktrace.NewProvider(sdktrace.WithResource(resource))
+```
+
+Or more simply:
+
+```go
+tp := sdktrace.NewProvider(sdktrace.MustDetectResource(awsdetector.EC2, gcpdetector.GCE, gcpdetector.GKE)) // or DetectResource (see below)
+```
+
+In the case of both `WithResource` & `DetectResource` being supplied, the detected resource will be merged with the supplied resource (with the supplied resource taking precedence).
+
+### Components
+
+#### Detector
+
+The Detector interface will simply return a Resource:
+
+```go
+type Detector interface {
+    Detect(ctx context.Context) (*Resource, error)
+}
+```
+
+If a detector is not able to detect a resource, it must return an uninitialized resource such that the result of each call to `Detect` can be merged.
+
+#### Provider
+
+A Provider will have a function to add detectors & to retrieve the auto-detected reosurce information:
+
+```go
+type Provider struct {
```

Comment:
Seems like Provider is just a composite Detector. Fewer concepts is better.

Reply:
I think @yurishkuro is correct here, and it was alluded to in the duplicate examples you gave above.

If we just went with the second approach:

```go
￼tp := sdktrace.NewProvider(sdktrace.MustDetectResource(awsdetector.EC2, gcpdetector.GCE, gcpdetector.GKE)) // or DetectResource
￼```

There would not be a need to add another provider.

Reply:
For the most part I agree, so I've removed this from the otep.

One disadvantage of removing this is that for each trace / metric provider you create, you will need to supply the list of detectors you want to run (rather than supplying a resource or provider) .. but that's not a huge issue, as you can just declare a list of detectors and re-use this (see the new example code)